test/test/jfr/JfrTests.java

```diff
@@ -74,4 +75,80 @@ public void parseMultiModeRecording(TestProcess p) throws Exception {
         Assert.isGreater(eventsCount.get("jdk.JavaMonitorEnter"), 50);
         Assert.isGreater(eventsCount.get("jdk.ObjectAllocationInNewTLAB"), 50);
     }
+
+    /**
+     * Test to validate time to safepoint profiling
+     *
+     * @param p The test process to profile with.
+     * @throws Exception Any exception thrown during profiling JFR output parsing.
+     */
+    @Test(mainClass = Ttsp.class)
+    public void onlyttsp(TestProcess p) throws Exception {
+        p.profile("-d 5 --ttsp -f %f.jfr");
+        if (containsSamplesOutsideWindow(p)) {
+            throw new RuntimeException("Expected no samples outside of ttsp window, but found some");
+        }
+    }
+
+    /**
+     * Test to validate time to safepoint profiling (recording the window only)
+     *
+     * @param p The test process to profile with.
+     * @throws Exception Any exception thrown during profiling JFR output parsing.
+     */
+    @Test(mainClass = Ttsp.class)
+    public void includettsp(TestProcess p) throws Exception {
+        p.profile("-d 5 --includettsp -f %f.jfr");
+        if (!containsSamplesOutsideWindow(p)) {
+            throw new RuntimeException("Expected to find samples outside of ttsp window, but did not find any");
+        }
+    }
+
+    /**
+     * Test using timebetween to validate time to safepoint profiling
+     *
+     * @param p The test process to profile with.
+     * @throws Exception Any exception thrown during profiling JFR output parsing.
+     */
+    @Test(mainClass = Ttsp.class)
+    public void timebetween(TestProcess p) throws Exception {
+        p.profile("-d 5 --timebetween SafepointSynchronize::begin RuntimeService::record_safepoint_synchronized -f %f.jfr");
+        if (!containsSamplesOutsideWindow(p)) {
+            throw new RuntimeException("Expected to find samples outside of ttsp window, but did not find any");
+        }
+    }
+
+    private boolean containsSamplesOutsideWindow(TestProcess p) throws Exception {
```

Comment:
Can be shortened and simplified a bit.
1. `events` list is redundant; we may collect profiler windows and execution samples in the first pass.
2. Profiler windows can be stored in a `TreeMap` as `startTime`->`endTime`. This will provide sorting and lookup for the closest window out of the box.
3. Search loop can be simplified with `stream().anyMatch()`.

Reply:
thanks!